habitat_baselines/common/rollout_storage.py

```diff
@@ -4,237 +4,196 @@
-        for sensor in observations:
-            self.observations[sensor][self.step + 1].copy_(
-                observations[sensor]
-            )
-        self.recurrent_hidden_states[self.step + 1].copy_(
-            recurrent_hidden_states
+        if not self.is_double_buffered:
+            assert buffer_index == 0
+
+        next_step = dict(
+            observations=observations,
+            recurrent_hidden_states=recurrent_hidden_states,
+            prev_actions=actions,
+            masks=masks,
         )
-        self.actions[self.step].copy_(actions)
-        self.prev_actions[self.step + 1].copy_(actions)
-        self.action_log_probs[self.step].copy_(action_log_probs)
-        self.value_preds[self.step].copy_(value_preds)
-        self.rewards[self.step].copy_(rewards)
-        self.masks[self.step + 1].copy_(masks)
 
-        self.step = self.step + 1
+        current_step = dict(
+            actions=actions,
+            action_log_probs=action_log_probs,
+            value_preds=value_preds,
+            rewards=rewards,
+        )
 
-    def after_update(self):
-        for sensor in self.observations:
-            self.observations[sensor][0].copy_(
-                self.observations[sensor][self.step]
-            )
+        next_step = {k: v for k, v in next_step.items() if v is not None}
+        current_step = {k: v for k, v in current_step.items() if v is not None}
 
-        self.recurrent_hidden_states[0].copy_(
-            self.recurrent_hidden_states[self.step]
+        env_slice = slice(
+            int(buffer_index * self._num_envs / self._nbuffers),
+            int((buffer_index + 1) * self._num_envs / self._nbuffers),
         )
-        self.masks[0].copy_(self.masks[self.step])
-        self.prev_actions[0].copy_(self.prev_actions[self.step])
-        self.step = 0
+
+        if len(next_step) > 0:
+            self.buffers.set(
+                (self.steps[buffer_index] + 1, env_slice),
+                next_step,
+                strict=False,
+            )
+
+        if len(current_step) > 0:
+            self.buffers.set(
+                (self.steps[buffer_index], env_slice),
+                current_step,
+                strict=False,
+            )
+
+    def advance_rollout(self, buffer_index: int = 0):
+        self.steps[buffer_index] += 1
+
+    def after_update(self):
+        self.buffers[0] = self.buffers[self.step]
+
+        self.steps = [0 for _ in self.steps]
 
     def compute_returns(self, next_value, use_gae, gamma, tau):
         if use_gae:
-            self.value_preds[self.step] = next_value
+            self.buffers["value_preds"][self.step] = next_value
             gae = 0
             for step in reversed(range(self.step)):
                 delta = (
-                    self.rewards[step]
-                    + gamma * self.value_preds[step + 1] * self.masks[step + 1]
-                    - self.value_preds[step]
+                    self.buffers["rewards"][step]
```

Comment:
[no action needed] As this is `common/rollout_storage.py` not sure how common this calculation are for common RL rollout.

Reply:
It's pretty common.  We would need a different data structure for off policy.